aptos-move/block-executor/src/view.rs

```diff
@@ -811,14 +1159,123 @@ impl<'a, T: Transaction, S: TStateView<Key = T::Key>, X: Executable> LatestView<
         read_set
             .iter()
             .filter(|key| !skip.contains(key))
-            .flat_map(|key| {
-                if let Some((value, Some(layout))) = unsync_map.fetch_data(key) {
-                    return self.does_value_need_exchange(
-                        &value,
-                        &layout,
-                        delayed_write_set_keys,
-                        key,
+            .flat_map(|key| match unsync_map.fetch_data(key) {
+                Some(ValueWithLayout::Exchanged(value, Some(layout))) => self
+                    .does_value_need_exchange(value.as_ref(), &layout, delayed_write_set_keys, key),
+                Some(ValueWithLayout::Exchanged(_, None)) => None,
+                Some(ValueWithLayout::RawFromStorage(_)) => Some(Err(code_invariant_error(
+                    "Cannot exchange value that was not exchanged before",
+                ))),
+                None => None,
+            })
+            .collect()
+    }
+
+    fn get_group_reads_needing_exchange_parallel(
+        &self,
+        parallel_state: &ParallelState<'a, T, X>,
+        delayed_write_set_keys: &HashSet<T::Identifier>,
+        skip: &HashSet<T::Key>,
+    ) -> Result<BTreeMap<T::Key, (T::Value, u64)>, PanicError> {
+        parallel_state
+            .captured_reads
+            .borrow()
+            .get_group_read_values_with_delayed_fields(skip)
+            .map(|(k, v)| (k.clone(), v.clone()))
+            .flat_map(|(key, group_read)| {
+                let GroupRead { inner_reads, .. } = group_read;
+                let mut resources_needing_delayed_field_exchange = false;
+                for data_read in inner_reads.values() {
+                    if let DataRead::Versioned(_version, value, Some(layout)) = data_read {
+                        if let Some(bytes) = value.bytes() {
+                            let identifiers_in_read = self
+                                .extract_identifiers_from_value(bytes, layout.as_ref())
+                                .unwrap();
+                            if !delayed_write_set_keys.is_disjoint(&identifiers_in_read) {
+                                // TODO[agg_v2](optimize): Is it possible to avoid clones here?
+                                resources_needing_delayed_field_exchange = true;
+                                break;
+                            }
+                        }
+                    }
+                }
+                if !resources_needing_delayed_field_exchange {
+                    return None;
+                }
+
+                // TODO[agg_v2](fix): Is it guaranteed that metadata is not None?
```

Comment:
I think if it's none it would have to correspond to a deletion, but then we wouldn't be able to read anything to replace out of it?
So if it's None, we should return some invariant error.

Reply:
Returning code invariant error here then.